### ISSUE TITLE
TxBoundary for cats.effect.IO

### ISSUE THREAD

**Issue by christobill** [scalikejdbc/scalikejdbc#1064]
IO monads are popular in the scala community
`cats.effect.IO` / `monix.eval.Task` / `scalaz.zio.ZIO`

It is fairly easy to provide a `TxBoundary` for `cats.effect.IO`
(and for other IO monad too)

We could then have a `ioLocalTx` (similar to `futureLocalTx`)

It would be a good alternative to the current solution to wrap
the `futureLocalTx`:
```
def ioLocalTx[A](execution: DBSession => IO[A])(implicit executionContext: ExecutionContext): IO[A] =
  IO.fromFuture(IO(DB.futureLocalTx(execution.andThen(_.unsafeToFuture()))))
```

It could be in a different `SubProject` if needed

**Comment by seratch** [comment]
Hello, @christobill 
Thanks for proposing this! 

>We could then have a ioLocalTx (similar to futureLocalTx)

I believe it's not necessary to add new methods to support those IO-based `TxBoundary` as you can give it to the existing `localTx` method (as an implicit param). In the case of `futureLocalTx`, I accepted the idea at the time but I don't have any plans to add more methods like that.

>It could be in a different SubProject if needed

Talking about ways to support the IO types, I would like to avoid having them as a part of the core library. The main reason is to keep the core library to have a quite limited number of dependencies the same as before.

If we cover those foundation libraries, the only option would be to have more sub-modules. One major concern I'm having about the idea is that, if some of them don't sometimes keep backward-compatibilities (either source-compatibility or binary-compatibility) among their versions, the release operations will be very complex like having their major version as part of artifactId - say, `scalikejdbc-cats-2.1_2.13`.

Long story short, I'm sorry to say this but, as the founder/main developer of this project, I'm not pretty much motivated to become the maintainer of those (for a long time from now on) yet.

I believe you can publish your own library that realizes the idea described here first. As said earlier, the`localTx` method accepts your types as an implicit param. So, there is no obstacle to build your library. If many people desire to have it as part of this library, let's revisit this and I will consider merging the changes to scalikejdbc project.

It would be greatly appreciated if I could get your understanding about this.

**Comment by christobill** [comment]
Hello @seratch 
Thanks for your quick and detailed answer!

> I believe it's not necessary to add new methods to support those IO-based TxBoundary as you can give it to the existing localTx method (as an implicit param)

It is true that `TxBoundary[T]` is a typeclass of which we can easily create extend :+1: . However working with `scalikejdbc` and `IO` can sometimes be a bit tricky. Mostly because `IO` is lazily evaluated and the evaluation happens after the connection is closed resulting in this:
```
[error] s.ActiveSession - Failed preparing the statement (Reason: Connection is null.):
  ()
```
(I can work on some minimal reproduction of the problem if you are interested)

We succeeded eventually with `futureLocalTx` :pray: (and we can probably do it in a simpler way, like you said and how I did in the PR, by "give it to the existing localTx method (as an implicit param)")

But the lazy execution problem for `readOnly` remains, because it has no `TxBoundary[T]`. It is logical since reading doesn't need a transaction, but it results in the error stated earlier

> The main reason is to keep the core library to have a quite limited number of dependencies the same as before

I totally understand your point about dependencies and maintainability. What if I could provide a simple implementation of an IO Monad for the tests, just to ensure lazy evaluations are handled by `scalikejdbc`? This would require no extra dependencies. I have a PR on the way regarding this matter

Above all I would like to make it easier for IO monads (`IO`,`ZIO`,`Task`) users to use `scalikejdbc`, either by this issue, either by adding documentation

**Comment by seratch** [comment]
@christobill 
>What if I could provide a simple implementation of an IO Monad for the tests, just to ensure lazy evaluations are handled by scalikejdbc? This would require no extra dependencies. I have a PR on the way regarding this matter

Cool, I'm happy with this idea :+1: 

>But the lazy execution problem for readOnly remains, because it has no TxBoundary[T]. It is logical since reading doesn't need a transaction, but it results in the error stated earlier

This is also true. Do you already have an idea to address this in mind? Probably, having `fReadOnly`, `fAutoCommit` etc? I'm fine with this as it won't bring any backward-incompatibilities.

**Comment by seratch** [comment]
Fixed by https://github.com/scalikejdbc/scalikejdbc.github.io/issues/87
